src/modules/DLGTE/OlympusGovDelegation.sol

```diff
@@ -308,69 +335,156 @@ contract OlympusGovDelegation is DLGTEv1 {
+            revert DLGTE_TooManyDelegates();
+        }
+
+        // Increase the delegation record amount
+        acctDelegatedAmounts.set(delegate, existingAmount + delegatedAmount);
+
+        // Delegate gOhm to the new escrow.
+        DelegateEscrow delegateEscrow = delegateEscrowFactory.create(delegate);
+        _gOHM.safeApprove(address(delegateEscrow), delegatedAmount);
+        delegateEscrow.delegate(onBehalfOf, delegatedAmount);
+        emit DelegationApplied(onBehalfOf, delegate, int256(delegatedAmount));
+    }
+
+    function _autoRescindDelegations(
+        address onBehalfOf,
+        uint256 requestedUndelegatedBalance,
+        AccountState storage aState,
+        uint256 totalAccountGOhm
+    ) private returns (uint256 actualUndelegatedBalance) {
+        actualUndelegatedBalance = totalAccountGOhm - aState.delegatedGOhm;
+
+        // Nothing to do if the undelegated balance is already greater than the requested amount
+        if (actualUndelegatedBalance >= requestedUndelegatedBalance)
+            return actualUndelegatedBalance;
+
+        // EnumerableMap internals are used here for gas efficiency.
+        // Deleting keys from the EnumerableMap changes the order (swap and pop) and size
+        // So take an upfront in-memory copy of the delegateAddrs keys to iterate over first.
+        EnumerableMap.AddressToUintMap storage acctDelegatedAmounts = aState.delegatedAmounts;
+        bytes32[] memory delegateAddrs = acctDelegatedAmounts._inner._keys._inner._values;
+        bytes32 delegateAddr;
+        uint256 delegatedBalance;
+        uint256 rescindAmount;
+        for (uint256 i; i < delegateAddrs.length; ++i) {
+            delegateAddr = delegateAddrs[i];
+            delegatedBalance = uint256(acctDelegatedAmounts._inner._values[delegateAddr]);
+
+            // Cap the amount to rescind for this delegate by the remaining required to get to the
+            // requested undelegated balance
+            rescindAmount = requestedUndelegatedBalance - actualUndelegatedBalance;
+            rescindAmount = delegatedBalance < rescindAmount ? delegatedBalance : rescindAmount;
+
+            _rescindDelegation(
+                onBehalfOf,
+                address(uint160(uint256(delegateAddr))),
+                delegatedBalance,
+                rescindAmount,
+                acctDelegatedAmounts
+            );
+
+            actualUndelegatedBalance += rescindAmount;
+
+            // Reached the requested undelegated balance
+            if (actualUndelegatedBalance == requestedUndelegatedBalance) break;
+        }
+
+        // Update state for the delegated amount of gOHM for this account
+        aState.delegatedGOhm = (totalAccountGOhm - actualUndelegatedBalance).encodeUInt112();
 
-        emit DelegationApplied(onBehalfOf, delegationRequest.delegate, delegatedDelta);
+        // May not have undelegated the full requested amount - left up to the calling policy on how to handle this gap
```

Comment:
I'm trying to understand in what scenario this would happen, and the only one that comes to mind is if the `requestedUndelegatedBalance` is greater than what was originally delegated. This seems to be the case in the `test_rescindDelegations_multiDelegatesLessThanEnough()` test function, where the `requestedUndelegatedBalance` is 520e18 vs an actual delegation of 400e18. If that's the case, why not catch that and revert? Otherwise, it seems like the exact `requestedUndelegatedBalance` value is possible, given this:

```
            actualUndelegatedBalance += rescindAmount;

            // Reached the requested undelegated balance
            if (actualUndelegatedBalance == requestedUndelegatedBalance) break;
```

Reply:
Trying to future proof. It could be that a future policy attempts to undelegate some amount (eg uint256 max) or some other optimistic amount, but cannot. 

If you revert here then that can't be supported in future. If you handle this way it's up to the policy to define the behaviour if it doesn't undelegate enough